File: easybluetooth/src/main/java/me/preetham/easybluetooth/BluetoothSpp.java
package me.preetham.easybluetooth;

import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothSocket;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.UUID;

import me.preetham.easybluetooth.enums.Mode;
import me.preetham.easybluetooth.interfaces.SppConnectionListener;
import me.preetham.easybluetooth.interfaces.SppDataListener;

public class BluetoothSpp {
    private static final String TAG = BluetoothSpp.class.getSimpleName();

    private UUID uuid;

    private SppDataListener dataListener;
    private SppConnectionListener connectionListener;

    private Mode mode;
    private int bufferSize;
    private boolean isRunning = false;

    private BluetoothSocket bluetoothSocket;
    private InputStream inputStream;
    private OutputStream outputStream;

    private BluetoothDevice targetDevice;


    private BluetoothSpp(Builder builder) {
        this.uuid = builder.uuid;
        this.dataListener = builder.dataListener;
        this.connectionListener = builder.connectionListener;
        this.mode = builder.mode;
        this.bufferSize = builder.bufferSize;
    }

    public static class Builder {
        private UUID uuid = UUID.fromString("00001101-0000-1000-8000-00805F9B34FB");

        private SppDataListener dataListener = null;
        private SppConnectionListener connectionListener = null;

        private Mode mode = Mode.INSECURE;
        private int bufferSize = 1024;

        public BluetoothSpp build() {
            return new BluetoothSpp(this);
        }

        public Builder setBufferSize(int bufferSize) {
            this.bufferSize = bufferSize;
            return this;
        }

        public Builder setConnectionListener(SppConnectionListener connectionListener) {
            this.connectionListener = connectionListener;
            return this;
        }

        public Builder setDataListener(SppDataListener dataListener) {
            this.dataListener = dataListener;
            return this;
        }

        public Builder setMode(Mode mode) {
            this.mode = mode;
            return this;
        }

        public Builder setUuid(String uuid) {
            this.uuid = UUID.fromString(uuid);
            return this;
        }

        public Builder setUuid(UUID uuid) {
            this.uuid = uuid;
            return this;
        }
    }

    private void start(final BluetoothSocket bluetoothSocket) {
        try {
            inputStream = bluetoothSocket.getInputStream();
            outputStream = bluetoothSocket.getOutputStream();

            if (connectionListener != null) {
                connectionListener.onConnectionSuccess();
                isRunning = true;
            }
        } catch (IOException e) {
            if (connectionListener != null) {
                connectionListener.onConnectionLost();
            }
            return;
        }

        Thread spp_thread = new Thread() {
            int bytes_read = -1;
            byte[] byte_data = new byte[bufferSize];

            @Override
            public void run() {
                while (isRunning) {
                    try {
                        bytes_read = inputStream.read(byte_data);
                        if (dataListener != null) {
                            dataListener.onDataReceived(byte_data, bytes_read);
                        }
                    } catch (IOException e) {
                        if (connectionListener != null) {
                            connectionListener.onConnectionLost();
                        }
                        break;
                    }
                }
            }
        };

        spp_thread.start();
    }

    public void connect(BluetoothDevice targetDevice) {
        try {
            switch (mode) {
                case SECURE:
                    bluetoothSocket = targetDevice.createRfcommSocketToServiceRecord(uuid);
                    break;
                case INSECURE:
                    bluetoothSocket = targetDevice.createInsecureRfcommSocketToServiceRecord(uuid);
                    break;
            }

            bluetoothSocket.connect();
            this.targetDevice = targetDevice;

            start(bluetoothSocket);
        } catch (IOException ex) {
            if (connectionListener != null) {
                connectionListener.onConnectionFailure();
            }
        }
    }

    public void disconnect() {
        isRunning = false;
        try {
            if (inputStream != null) {
                inputStream.close();
            }

            if (outputStream != null) {
                outputStream.close();
            }

            if (bluetoothSocket != null) {
                bluetoothSocket.close();
            }

            if (connectionListener != null) {
                connectionListener.onConnectionClosed();
            }
        } catch (IOException ignored) {

        }
    }

    public BluetoothDevice getTargetDevice() {
        return targetDevice;
    }

    public boolean isRunning() {
        return isRunning;
    }

    public void write(byte[] bytes) {
        try {
            outputStream.write(bytes);
            if (dataListener != null) {
                dataListener.onDataSent(bytes);
            }
        } catch (IOException e) {
            connectionListener.onConnectionLost();
        }
    }

    public void write(String data_string) {
        write(data_string.getBytes());
    }
}

File: easybluetooth/src/main/java/me/preetham/easybluetooth/BluetoothSpp.kt
package me.preetham.easybluetooth

import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothSocket
import me.preetham.easybluetooth.enums.Mode
import me.preetham.easybluetooth.interfaces.SppConnectionListener
import me.preetham.easybluetooth.interfaces.SppDataListener
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.util.*

open class BluetoothSpp private constructor(builder: Builder) {

    private val uuid: UUID

    private val dataListener: SppDataListener?
    private val connectionListener: SppConnectionListener?

    private val mode: Mode
    private val bufferSize: Int
    var isRunning = false
        private set

    private var bluetoothSocket: BluetoothSocket? = null
    private var inputStream: InputStream? = null
    private var outputStream: OutputStream? = null

    var targetDevice: BluetoothDevice? = null
        private set


    init {
        this.uuid = builder.uuid
        this.dataListener = builder.dataListener
        this.connectionListener = builder.connectionListener
        this.mode = builder.mode
        this.bufferSize = builder.bufferSize
    }

    class Builder {
        var uuid: UUID = UUID.fromString("00001101-0000-1000-8000-00805F9B34FB")
            private set

        var dataListener: SppDataListener? = null
            private set

        var connectionListener: SppConnectionListener? = null
            private set

        var mode = Mode.INSECURE
            private set

        var bufferSize = 1024
            private set

        fun build(): BluetoothSpp {
            return BluetoothSpp(this)
        }

        fun setBufferSize(bufferSize: Int): Builder = apply {
            this.bufferSize = bufferSize
        }

        fun setConnectionListener(connectionListener: SppConnectionListener): Builder = apply {
            this.connectionListener = connectionListener
        }

        fun setDataListener(dataListener: SppDataListener): Builder = apply {
            this.dataListener = dataListener
        }

        fun setMode(mode: Mode): Builder = apply {
            this.mode = mode
        }

        fun setUuid(uuid: String): Builder = apply {
            this.uuid = UUID.fromString(uuid)
        }

        fun setUuid(uuid: UUID): Builder = apply {
            this.uuid = uuid
        }
    }

    private fun start(bluetoothSocket: BluetoothSocket) {
        try {
            inputStream = bluetoothSocket.inputStream
            outputStream = bluetoothSocket.outputStream

            connectionListener?.onConnectionSuccess()
            isRunning = true
        } catch (e: IOException) {
            connectionListener?.onConnectionLost()
            return
        }

        val sppThread = object : Thread() {
            var bytes_read = -1
            var byte_data = ByteArray(bufferSize)

            override fun run() {
                while (isRunning) {
                    try {
                        bytes_read = inputStream!!.read(byte_data)
                        dataListener?.onDataReceived(byte_data, bytes_read)
                    } catch (e: IOException) {
                        connectionListener?.onConnectionLost()
                        break
                    }
                }
            }
        }

        sppThread.start()
    }

    fun connect(targetDevice: BluetoothDevice) {
        try {
            bluetoothSocket = when (mode) {
                Mode.SECURE -> targetDevice.createRfcommSocketToServiceRecord(uuid)
                Mode.INSECURE -> targetDevice.createInsecureRfcommSocketToServiceRecord(uuid)
            }

            bluetoothSocket!!.connect()
            this.targetDevice = targetDevice

            start(bluetoothSocket!!)
        } catch (ex: IOException) {
            connectionListener?.onConnectionFailure()
        }
    }

    fun disconnect() {
        isRunning = false
        try {
            inputStream?.close()
            outputStream?.close()
            bluetoothSocket?.close()
            connectionListener?.onConnectionClosed()
        } catch (ignored: IOException) {

        }
    }

    fun write(bytes: ByteArray) {
        try {
            outputStream!!.write(bytes)
            dataListener?.onDataSent(bytes)
        } catch (e: IOException) {
            connectionListener!!.onConnectionLost()
        }

    }

    fun write(data_string: String) {
        write(data_string.toByteArray())
    }

    companion object {
        private val TAG = BluetoothSpp::class.java.simpleName
    }
}
